docs/architecture/adr-015-ibc-packet-receiver.md

```diff
@@ -0,0 +1,247 @@
+# ADR 015: IBC Packet Receiver
+
+## Changelog
+
+- 2019 Oct 22: Initial Draft
+
+## Context
+ 
+[ICS 26 - Routing Module](https://github.com/cosmos/ics/tree/master/spec/ics-026-routing-module) defines function [`handlePacketRecv`](https://github.com/cosmos/ics/tree/master/spec/ics-026-routing-module#packet-relay).
+
+`handlePacketRecv` executes per-module `onRecvPacket` callbacks, verifies the
+packet merkle proof, and pushes the acknowledgement bytes, if present, to the IBC
+channel `Keeper` state (ICS04).
+
+`handlePacketAcknowledgement` executes per-module `onAcknowledgementPacket`
+callbacks, and verifies the acknowledgement commitment proof.
+
+`handlePacketTimeout` and `handlePacketTimeoutOnClose` executes per-module
+`onTimeoutPacket` callbacks, and verifies the timeout proof.
+
+The mechanism is similar to the transaction handling logic in `baseapp`. After
+authentication, the handler is executed, and the authentication state change
+must be committed regardless of the result of the handler execution.
+
+`handlePacketRecv` also requires acknowledgement writing which has to be done
+after the handler execution and also must be commited regardless of the result of
+the handler execution.
+
```

Comment:
The context doesn't really define any problem...at least I can't see one immediately. I would state the problem clearly (purpose of the ADR) and how it relates to the ante-handler.